File: 0160-intersection-of-two-linked-lists/0160-intersection-of-two-linked-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def getIntersectionNode(self, headA: ListNode, headB: ListNode) -> Optional[ListNode]:
        if not headA and not headB:
            return null
        
        rootA, rootB = headA, headB
        
        while rootA != rootB:
            rootA = rootA.next if rootA else headB
            rootB = rootB.next if rootB else headA
        
        return rootA       

File: 0160-intersection-of-two-linked-lists/0160-intersection-of-two-linked-lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
        if (!headA || !headB) {
            return nullptr;
        }
        
        ListNode* rootA = headA;
        ListNode* rootB = headB;
        
        while (rootA != rootB) {
            rootA = rootA ? rootA->next : headB;
            rootB = rootB ? rootB->next : headA;
        }
        
        return rootA;
    }
};